experimental/iterators/include/iterators/Dialect/Iterators/IR/IteratorsOps.td

```diff
@@ -432,6 +432,46 @@ def Iterators_ValueToStreamOp : Iterators_Op<"value_to_stream",
   }];
 }
 
+class NonemptyVariadic<Type type> : Variadic<type> { let minSize = 1; }
+
+def Iterators_ZipOp : Iterators_Op<"zip",
+    [AllMatch<[[{::llvm::ArrayRef(::llvm::SmallVector<Type>(
+                    ::llvm::map_range($inputs.getTypes(),
+                      [](Type t) { return t.cast<StreamType>().getElementType(); }
+                    )))}],
+               [{$result.getType().cast<StreamType>().getElementType()
+                    .cast<::mlir::TupleType>().getTypes()}]],
+              "result stream must consist of tuples whose element types match "
+              "the element types of the input streams">,
+     DeclareOpInterfaceMethods<OpAsmOpInterface, ["getAsmResultNames"]>]> {
+  let summary = "Zips several streams into a stream of tuples.";
+  let description = [{
+    Reads one or more streams in lock step and produces a stream of tuples where
+    each struct constists of the elements of all input streams at the same
+    position of the streams.
+
```

Comment:
You want to describe the termination condition.

For instance LLVM also as zip_equal that asserts all streams have the same number of elements.

Reply:
OK, done.